File: ante.go
package main

import (
	`bytes`
	`fmt`
	`io/ioutil`
	`math/big`
	`os`
	`regexp`
	`strings`
	`unicode/utf8`
)

type Card struct {
	rank rune
	suit rune
}

type Ante struct {
	pc     int               // Program counter (index within ante.code)
	line   int               // Current line number.
	code   []Card            // Array of cards.
	vars   map[rune]*big.Int // Four registers hashed by suit.
	labels map[rune]int      // Labels for ante.pc to jump to.
	buffer []byte            // Buffer to collect UTF-8 character bytes.
}

func (ante *Ante) Initialize() *Ante {
	ante.labels = map[rune]int{}
	ante.vars = map[rune]*big.Int{
		'♦': big.NewInt(0),
		'♥': big.NewInt(0),
		'♠': big.NewInt(0),
		'♣': big.NewInt(0),
	}

	return ante
}

func (ante *Ante) Run(filename string) *Ante {
	program, err := ioutil.ReadFile(filename)
	if err != nil {
		ante.exception(fmt.Sprintf(`%q`, err))
	}

	ante.parse(program)

	for ante.pc < len(ante.code) {
		card := ante.code[ante.pc]
		ante.pc++

		switch card.rank {
		case 0:
			ante.newline(card)
		case 'K':
			ante.jump(card)
		case 'Q':
			continue
		case 'J':
			ante.dump(card, true)
		case 10:
			ante.dump(card, false)
		default:
			ante.assign(card)
		}
	}

	return ante
}

func (ante *Ante) parse(program []byte) *Ante {
	// Split program blob into lines.
	lines := bytes.Split(program, []byte("\n"))

	// Get rid of comments and whitespaces.
	comments := regexp.MustCompile(`#.*$`)
	whitespace := regexp.MustCompile(`^\s+|\s+$`)
	for i := 0; i < len(lines); i++ {
		lines[i] = comments.ReplaceAllLiteral(lines[i], []byte(``))
		lines[i] = whitespace.ReplaceAllLiteral(lines[i], []byte(``))
		//fmt.Printf("[%s]\n", lines[i])
	}

	// Turn source file into array of cards.
	re := regexp.MustCompile(`(10|[2-9JQKA])([♦♥♠♣])`)
	for i := 0; i < len(lines); i++ {
		// Line number cards have 0 rank.
		ante.code = append(ante.code, Card{0, rune(i + 1)})
		cards := re.FindAllSubmatch(lines[i], -1)
		for _, c := range cards {
			ante.code = append(ante.code, Card{bytes.Runes(c[1])[0], bytes.Runes(c[2])[0]})
		}
	}

	// A pass to convert ranks to integers and extract labels.
	for pc := 0; pc < len(ante.code); {
		card := ante.code[pc]
		pc++
		if card.rank == '1' {
			ante.code[pc-1].rank = 10
		} else if card.rank >= '2' && card.rank <= '9' {
			ante.code[pc-1].rank = card.rank - '0'
		} else if card.rank == 'Q' {
			queen := card.suit
			for pc < len(ante.code) && ante.code[pc].rank == 'Q' && ante.code[pc].suit == card.suit {
				queen += card.suit
				pc++
			}
			ante.labels[queen] = pc
		}
	}
	//fmt.Printf("%q\n", ante.code)
	//fmt.Printf("%q\n", ante.labels)

	return ante
}

func (ante *Ante) newline(card Card) *Ante {
	ante.line++

	return ante
}

func (ante *Ante) jump(card Card) *Ante {
	//fmt.Printf("jump: %q, %d, %q\n", card, ante.pc, ante.labels)
	suit := card.suit
	for ante.pc < len(ante.code) && ante.code[ante.pc].rank == 'K' && ante.code[ante.pc].suit == card.suit {
		suit += card.suit
		ante.pc++
	}

	// if ante.vars[card.suit] != 0 ...
	if ante.vars[card.suit].Cmp(big.NewInt(0)) != 0 {
		if _, ok := ante.labels[suit]; ok {
			ante.pc = ante.labels[suit]
		} else {
			label := strings.Repeat(fmt.Sprintf(`Q%c`, card.suit), int(suit/card.suit))
			ante.exception(`can't find ` + label + ` to go`)
		}
	}

	return ante
}

func (ante *Ante) dump(card Card, char bool) *Ante {
	//fmt.Printf("dump %q\n", char)
	value := ante.vars[card.suit]
	if char {
		// if value < 0 || value > 255 ...
		if value.Cmp(big.NewInt(0)) == -1 || value.Cmp(big.NewInt(255)) == 1 {
			ante.exception(fmt.Sprintf(`character code %d is out of 0..255 range`, value))
		} else {
			// Collect the bytes till we have full UTF-8 character.
			// Once the character is built dump it and reset the buffer.
			ante.buffer = append(ante.buffer, byte(value.Int64()))
			if utf8.FullRune(ante.buffer) {
				fmt.Printf(`%s`, ante.buffer)
				ante.buffer = []byte{}
			}
		}
	} else {
		fmt.Print(value)
	}

	return ante
}

func (ante *Ante) assign(card Card) *Ante {
	//fmt.Printf("assign %q\n", card)
	operands := ante.remaining(card)

	return ante.expression(operands)
}

// Fetch the rest of the assignment expression.
func (ante *Ante) remaining(card Card) []Card {
	operands := []Card{card}
	for ante.pc < len(ante.code) {
		card = ante.code[ante.pc]
		if card.rank == 0 || card.rank == 'K' || card.rank == 'Q' || card.rank == 'J' {
			break
		}
		operands = append(operands, card)
		ante.pc++
	}

	//fmt.Printf("remaining: %q\n", operands)
	return operands
}

func (ante *Ante) expression(operands []Card) *Ante {
	initial := big.NewInt(int64(operands[0].rank))
	target := operands[0].suit

	// if initial == 'A' ...
	if initial.Cmp(big.NewInt('A')) == 0 {
		// initial = ante.vars[target] // <-- Wrong!!! We need local copy, not a pointer to register.
		initial.Set(ante.vars[target]) // <-- The right way to interpolate.
	}

	// Go through remaining operands.
	for _, card := range operands[1:] {
		rank := big.NewInt(int64(card.rank))
		suit := card.suit
		//fmt.Printf("rank: %d, suit: %c\n", rank, suit)

		// if rank == 'A' ...
		if rank.Cmp(big.NewInt('A')) == 0 {
			rank.Set(ante.vars[suit]) // <-- The right way to interpolate.
		}
		switch suit {
		case '♦':
			initial.Add(initial, rank)
		case '♥':
			initial.Mul(initial, rank)
		case '♠':
			initial.Sub(initial, rank)
		case '♣':
			if rank.Cmp(big.NewInt(0)) != 0 {
				initial.Div(initial, rank)
			} else {
				ante.exception(`division by zero`)
			}
		}
	}
	ante.vars[target] = initial
	//fmt.Printf("  %d %c => %d\n", ante.pc, target, initial)

	return ante
}

func (ante *Ante) exception(message string) {
	fmt.Printf("Ante exception: %s on line %d (pc:%d)\n", message, ante.line, ante.pc)
	os.Exit(1)
}

func main() {
	if len(os.Args) == 2 {
		new(Ante).Initialize().Run(os.Args[1])
	} else {
		fmt.Println(`usage: ante filename.ante`)
	}
}

File: ante.rs
// Copyright (c) 2014-2015 Michael Dvorkin
// Ante is an esoteric programming language where all you've got is a deck of cards.
//
// This is Ante implementation in Rust.

extern crate regex;
extern crate num;

use std::io::File;
use std::collections::HashMap;
use std::num::Zero;
use std::num::from_int;
use std::str::{is_utf8, from_utf8};
use std::char::{from_u32};
use std::os;
use regex::Regex;
use num::bigint::BigInt;

const ACE:      uint = 'A' as uint;
const JACK:     uint = 'J' as uint;
const QUEEN:    uint = 'Q' as uint;
const KING:     uint = 'K' as uint;
const DIAMONDS: uint = '♦' as uint;
const HEARTS:   uint = '♥' as uint;
const SPADES:   uint = '♠' as uint;
const CLUBS:    uint = '♣' as uint;

struct Card {
    rank: uint,
    suit: uint
}

struct Ante {
    pc:     uint,                   // Program counter (index within ante.code)
    line:   uint,                   // Current line number.
    code:   Vec<Card>,              // Array of cards.
    vars:   HashMap<uint, BigInt>,  // Four registers hashed by suit.
    labels: HashMap<uint, uint>,    // Labels for ante.pc to jump to.
    buffer: Vec<u8>                 // Buffer to collect UTF-8 character bytes.
}

impl Ante {
    fn new(filename: &str) -> Ante {
        let code = Ante::parse(filename);
        let labels = Ante::resolve(&code);
        let mut vars: HashMap<uint, BigInt> = HashMap::new();
        vars.insert(DIAMONDS, Ante::big(0));
        vars.insert(HEARTS, Ante::big(0));
        vars.insert(SPADES, Ante::big(0));
        vars.insert(CLUBS, Ante::big(0));

        Ante {
            pc:     0,
            line:   0,
            code:   code,
            vars:   vars,
            labels: labels,
            buffer: vec![]
        }
    }

    fn run(&mut self) -> &Ante {
        while self.pc < self.code.len() {
            let card = self.code[self.pc];
            self.pc += 1;
            match card.rank {
                0u    => self.newline(card),
                10u   => self.dump(card, false),
                JACK  => self.dump(card, true),
                QUEEN => continue,
                KING  => self.jump(card),
                _     => self.assign(card)
            };
        }
        self
    }

    // Turn source file into array of cards.
    fn parse(filename: &str) -> Vec<Card> {
        let mut file = match File::open(&Path::new(filename)) {
            // The `desc` field of `IoError` is a string that describes the error.
            Err(reason) => fail!("couldn't open {} ({})", filename, reason.desc),
            Ok(file) => file,
        };
        let program = file.read_to_string().unwrap();

        // Split program blob into lines getting rid of comments and whitespaces.
        let comments = Regex::new(r"#.*$").unwrap();
        let lines: Vec<String> = program.as_slice().lines().map( |line|
            comments.replace_all(line, "").as_slice().trim().to_string()
        ).collect();

        // Turn source file into array of cards. Each card a struct of rank and suit.
        let mut code: Vec<Card> = vec![];
        let card = Regex::new(r"(10|[2-9JQKA])([♦♥♠♣])").unwrap();
        for (i, line) in lines.iter().enumerate() {
            // Line number cards have zero rank.
            code.push(Card { rank: 0, suit: i + 1 });

            // Parse lines to turn them into array of cards.
            for caps in card.captures_iter(line.as_slice()) {
                let rank = caps.at(1).char_at(0);
                let suit = caps.at(2).char_at(0);
                let card = match rank {
                   '1'       => Card { rank: 10, suit: suit as uint },
                   '2'...'9' => Card { rank: rank as uint - 48, suit: suit as uint },
                   _         => Card { rank: rank as uint, suit: suit as uint }
                };
                code.push(card);
            }
        }
        code
    }

    // Extra pass to set up jump labels.
    fn resolve(code: &Vec<Card>) -> HashMap<uint, uint> {
        let mut pc = 0u;
        let mut labels: HashMap<uint, uint> = HashMap::new();

        while pc < code.len() - 1 {
            let card = code[pc];
            pc += 1;
            if card.rank == QUEEN {
                let mut queen = card.suit as uint;
                while pc < code.len() && code[pc].rank == QUEEN && code[pc].suit == card.suit {
                    queen += card.suit as uint;
                    pc += 1;
                }
                labels.insert(queen, pc);
            }
        }
        labels
    }

    fn newline(&mut self, card: Card) -> &Ante {
        self.line = card.suit as uint;
        self
    }

    fn jump(&mut self, card: Card) -> &Ante {
        let mut suit = card.suit as uint;
        while self.pc < self.code.len() && self.code[self.pc].rank == KING && self.code[self.pc].suit == card.suit {
            suit += card.suit as uint;
            self.pc += 1;
        }

        if !self.vars[card.suit].is_zero() {
            let label = suit as uint;
            if self.labels.contains_key(&label) {
                self.pc = self.labels[label];
            } else {
                let label = format!("Q{:c}", from_u32(card.suit as u32).unwrap());
                self.exception(format!("can't find {} to go to", label.repeat(suit / card.suit)).as_slice());
            }
        }
        self
    }

    fn assign(&mut self, card: Card) -> &Ante {
        let operands = self.remaining(card);
        self.expression(operands)
    }

    fn dump(&mut self, card: Card, as_character: bool) -> &Ante {
        let value = self.vars[card.suit].clone();
        if as_character {
            if value < Ante::big(0) || value > Ante::big(255) {
                self.exception(format!("character code {} is not in 0..255 range", value).as_slice());
            } else {
                // Collect the bytes till we have full UTF-8 character.
                // Once the character is built dump it and reset the buffer.
                self.buffer.push(value.to_u8().unwrap());
                if is_utf8(self.buffer.as_slice()) {
                    print!("{}", from_utf8(self.buffer.as_slice()).unwrap());
                    self.buffer = vec![];
                }
            }
        } else {
            print!("{}", value);
        }
        self
    }

    fn remaining(&mut self, card: Card) -> Vec<Card> {
        let mut operands: Vec<Card> = vec![card];

        while self.pc < self.code.len() {
            let card = self.code[self.pc];
            if card.rank == 0 || card.rank == KING || card.rank == QUEEN || card.rank == JACK {
                break;
            }
            operands.push(card);
            self.pc += 1;
        }
        operands
    }

    fn expression(&mut self, operands: Vec<Card>) -> &Ante {
        let mut initial: BigInt = Ante::big(operands[0].rank as int);
        let target = operands[0].suit;

        if initial.to_uint().unwrap() == ACE {
            initial = self.vars[target].clone();
        }

        for i in range(1, operands.len()) {
            let mut rank = Ante::big(operands[i].rank as int);
            let suit = operands[i].suit;

            if rank.to_uint().unwrap() == ACE {
                rank = self.vars[suit].clone();
            }

            match suit {
                DIAMONDS => { initial = initial.add(&rank) },
                HEARTS   => { initial = initial.mul(&rank) },
                SPADES   => { initial = initial.sub(&rank) },
                CLUBS    => if !rank.is_zero() {
                                initial = initial.div(&rank);
                            } else {
                                self.exception("division by zero");
                            },
                _        => continue
            }
        }

        *self.vars.get_mut(&target) = initial;
        self
    }

    fn big(n: int) -> BigInt {
        from_int::<BigInt>(n).unwrap()
    }

    // NOTE: fail! got renamed to panic!
    fn exception(&self, message: &str) {
        fail!("Ante exception: {} on line {} (pc:{})\n", message, self.line, self.pc)
    }
}


fn main() {
    if os::args().len() == 2 {
        Ante::new(os::args()[1].as_slice()).run();
    } else {
        println!("usage: ante filename.ante");
    }
}
